application/models/Attribute.php

```
@@ -244,30 +244,80 @@ public function get_found_rows($search)
 		return $this->search($search)->num_rows();
 	}
 
+	public function convert_definition_type($definition_id,$from_type,$to_type)
+	{
+		if($from_type === TEXT)
+		{
+			//From TEXT to DATETIME
+			if($to_type === DATETIME)
+			{
+				$this->db->trans_start();
+				
+				$query = 'UPDATE ospos_attribute_values ';
+				$query .= 'INNER JOIN ospos_attribute_links ';
+				$query .= 'ON ospos_attribute_values.attribute_id = ospos_attribute_links.attribute_id ';
+				$query .= 'SET attribute_datetime = attribute_value, ';
+				$query .= 'attribute_value = NULL ';
+				$query .= 'WHERE definition_id = '.$definition_id;
+				$this->db->query($query);
+				
+				$this->db->trans_complete();
```

What if the data can't be converted to datetime format? Maybe check how the code behaves in that case, do we get a notification popup to notify the user that the update failed?

You might want to update the $success variable to make that work then here.

---

Yeah, we need to hash that out.  Currently with improperly formatted data MySQL puts 0000-00-00 00:00:00.  I posed this question in the issue thread, but didn't get any responses.  We should decide on what the behavior should be.  We can leave it as it, or abort the change and kick back an error.  I'm not sure how we could do the latter unless I do a query prior to this and search the results using php because the query doesn't kick back an error with junk data.  It just quietly inserts the zeros.  Ideas?

---

I guess in case we do something like inserting a default value to replace unparseable dates, we should warn the user first. Otherwise I would just throw an error saying that values cannort be converted and that this needs to be fixed in some way.

---

added $success to that query, however like I said before, MySQL doesn't kick an error on junk data, so I may need to write a query prior to this using Query Builder to pull the data and examine it.  I'll refactor that into another function since we may want to have data validity checking on other attribute conversions in the future.  I think to simplify it, I will just kick back an array of item_ids with the improperly formatted data and have that output to the error log with a toast notification that the conversion failed and that they should check the error log.

---

@jekkos @daN4cat OK, I've added code to check the formatting of the TEXT on conversion to DATETIME and both output each line that isn't either YYYY-MM-DD or YYYY-MM-DD HH:MM:SS and then throw an error in the toast notification.

One problem that we have is that currently the application prevents the ':' symbol in a TEXT attribute but the migration allows it.  This means that if a user needs to correct the formatting on one of the attributes they won't be able to enter a YYYY-MM-DD HH:MM:SS format.  Is there a reason we are blocking ':'?

---

Well I think I want to drop the time part for datetime attribute, as I think it's usage is limited. Most of the time people use will use it for expiry date tracking or the like. So not sure if we need a complete timestamp in there?

---

@jekkos For our use, we don't need a full timestamp it as we are dealing with the publication date.  So then would you change the database structure to have the attribute_datetime be a DATE in MySQL instead of DATETIME?  Also, does it make sense then to refactor our code and the name of the column to be DATE and attribute_date respectively?  Let me know what you end of deciding on just storing the date and I will adjust the regex accordingly.  The only scenario in which I could forsee the need for a full DATETIME is if we expanded attributes in the future to include Inventory attributes. In that case it could be useful to have a datetime which tells us the hour and minute an item was received, quantity was changed, report was run, etc.

Alternately we could just keep the database as is for future compatibility and ignore the hour, minute and second on the software end.  MySQL will accept YYYY-MM-DD in a datetime without the time portion.  It just stores zeros in the time portion.